### PR TITLE
rfcs: extend dropout with offset, s64 and host_scalars

### PR DESCRIPTION
**Pull request by mgouicem**
This RFC proposes to extend dropout attribute to support #3233 .
[Link to rendered document](https://github.com/uxlfoundation/oneDNN/blob/fdd27a96d860d44f133bf7173694f33062944d68/rfcs/20250923-dropout-postop-extension/README.md)
POC is WIP in #4031

### CODE REVIEW

File: rfcs/20250923-dropout-postop-extension/README.md [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+/// @param mask_desc Output mask memory descriptor
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t DNNL_API dnnl_primitive_attr_set_dropout_v2(
+        dnnl_primitive_attr_t attr, const_dnnl_memory_desc_t mask_desc, 
+        dnnl_data_type_t seed_dt, int use_offset, int use_host_scalars);
+```
+
+The symbol `dnnl_primitive_attr_set_dropout` will just become a call
+to `dnnl_primitive_attr_set_dropout_v2` with `seed_dt=f32`,
```
Comment (ElaineBao) [inline rfcs/20250923-dropout-postop-extension/README.md:161]: I guess the default seed dtype is s32?

File: rfcs/20250923-dropout-postop-extension/README.md [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+
+## Example
+
+```c++
+using memory;
+
+primitive_attr attr;
+attr.set_dropout(memory::desc(), // mask will not be written to memory
+        s64, true); // probability/seed/offset will be passed as s64 host scalars.
```
Comment (ElaineBao) [inline rfcs/20250923-dropout-postop-extension/README.md:192]: probability's dtype should be f32 I think, which is not defined by this argument

File: rfcs/20250923-dropout-postop-extension/README.md [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+### Extend dropout attribute support to other primitives
+
+Skipping dropout mask writing on forward is only useful if both
+primitives around the dropout node support the dropout attribute.
+Currently, the dropout attribute is only supported by Matmul
+primitive. However, to support sdpa pattern for example (as described
+[here](https://github.com/uxlfoundation/oneDNN/pull/3233), dropout
+happens after softmax and before matmul on forward. This implies that
+dropout attribute would need to be supported by softmax to support
+this pattern.
```
Comment (ElaineBao) [inline rfcs/20250923-dropout-postop-extension/README.md:225]: To support sdpa backward pattern, eltwise-exp also needs to support dropout attribute, as softmax is decomposed to subtract+exp in backward, see the forward and backward graphs in [proposal 2.B of Supporting SDPA Training in Graph API](https://github.com/uxlfoundation/oneDNN/blob/yixin/rfc/sdpa-training/rfcs/20250428-graph-api-sdpa-training/README.md#proposal-2b-extend-existing-softmax-operation).

File: rfcs/20250923-dropout-postop-extension/README.md (hunk 2 above)
Comment (dzarukin) [reply rfcs/20250923-dropout-postop-extension/README.md:192]: Besides, there's not enough args passed to the call, as there's two specified while API defines three.

File: rfcs/20250923-dropout-postop-extension/README.md [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+`use_offset=0` and `use_host_scalars=0`.
+
+On the C++ API side, set_dropout will now call
+`dnnl_primitive_attr_set_dropout_v2` and there would be a new overload
+with new parameters as follow:
+
+```c++
+    /// Sets dropout probability.
+    ///
+    /// @param mask_desc Output memory descriptor of a dropout mask.
```
Comment (dzarukin) [inline rfcs/20250923-dropout-postop-extension/README.md:171]: It's rather an implementation defined comment, but the behavior about empty md should be specified as well, I think, for better clarity.

File: rfcs/20250923-dropout-postop-extension/README.md (hunk 1 above)
Comment (dzarukin) [reply rfcs/20250923-dropout-postop-extension/README.md:161]: Please check some more places below, this typo is not in a single spot, including proposed API.

**Review by dzarukin — APPROVED** [review]

**Comment by openmodelzooci** [comment]
Can one of the admins verify this patch?

**Review by vpirogov — APPROVED** [review]
